HW2/P2/parallel_vector.pyx

```diff
@@ -80,12 +80,37 @@ cpdef move_data_fine_grained(np.int32_t[:] counts,
    # data movement.  Be sure to avoid deadlock, and double-locking.
    ##########
    with nogil:
-       for r in range(repeat):
+       for r in prange(repeat, num_threads=4):
            for idx in range(src.shape[0]):
+               if src[idx] < dest[idx]: 
+                  acquire(&locks[src[idx]])
+                  acquire(&locks[dest[idx]])
+                  src_dst_comp = 1
+               elif src[idx] > dest[idx]:
+                  acquire(&locks[dest[idx]])
+                  acquire(&locks[src[idx]])
+                  src_dst_comp = 2
+               else:
+                # they are equal, acquire only once 
+                  acquire(&locks[src[idx]])
+                  src_dst_comp = 0
+
+               # do the swap
                if counts[src[idx]] > 0:
                    counts[dest[idx]] += 1
                    counts[src[idx]] -= 1
 
+               # release the locks, reverse order
+               if src_dst_comp == 0:
+                    release(&locks[src[idx]])
+               elif src_dst_comp == 1:
+                    release(&locks[dest[idx]])
+                    release(&locks[src[idx]])
+               else:
+                    release(&locks[src[idx]])
+                    release(&locks[dest[idx]])
```

Comment:
You could consolidate this large if/else block and avoid rewriting the same code by just using something along the lines of 'aquire(&(locks[min(src[idx],dest[idx])]))' etc. But other than that, this works great!